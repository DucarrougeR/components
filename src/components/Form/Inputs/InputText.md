```js noeditor
const StatusAndResources = require('../../../../styleguide_components/StatusAndResources').StatusAndResources;

<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/2MG6DoSjk4IaLnjjFCnKFf/Forms"
  githubURL="https://github.com/looker/relens/blob/master/src/components/Form/Inputs/InputText.tsx"
  feedbackTitle="InputText Component Feedback" />
```

### Defalut InputText

The `<InputText />` component renders a single input element on the page, with no accompanying label. It is generally used to construct higher-order components like the `<FieldText />`. If you are building a form, you probably want to use `<FieldText />` instead as it provides label support.

### Name and ID

A name and ID can be specified in the `<InputText />` component. Names are important if the checkbox is used in the context of a form, in which case a name is required for the value of the input to be captured.

```js
  <InputText name="someName" id="someId" />
```

### Disabled Property

Use the disable property to make an input field uneditable.

```js
  <InputText disabled />
  <InputText value="Hello!" disabled />
```

### Placeholders

Placeholders are used to give a hint to the user of the expected value for the input.

```js
  <InputText placeholder="Enter your text here..." />
```

### Input Types

Four types of input text are supported: email, number, password, and text.

**Email**: Will validate that the inputted value is an email address when submitted

**Number**: Forces a numeric value

**Password**: Hides typed characters

**Text**: Regular text input

_Note: Because components in Lens do not store state, if a component has a value specified (as they do in this example), it cannot be changed. In actual implementation, the user will need to define an onChange handler to update the value._

```js
  <InputText type="email" value="example@gmail.com"/>
  <InputText type="number" value="5" />
  <InputText type="password" value="password1234" />
  <InputText type="text" value="Regular old text" />
```

### ReadOnly Property

As the name suggests, `readOnly` makes the text uneditable.

```js
  <InputText value="You can't change me." readOnly />
```
