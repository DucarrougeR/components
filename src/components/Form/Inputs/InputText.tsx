// Typescript component boilerplate for generating meaningful declaration files.
import * as React from 'react'
import styled, { StyledComponentClass } from '../../../styled_components'
export { StyledComponentClass }
import { ThemeInterface } from '../../../themes'
export { ThemeInterface }
import { themeSpacing } from '../../../themes/theme_spacing'
import { FontRamp, fontSizes } from '../../../styles/font_sizes'
// End Typescript component boilerplate

export enum InputTextTypes {
  Email = 'email',
  Number = 'number',
  Password = 'password',
  Text = 'text'
}

export interface InputTextProps extends React.HTMLProps<HTMLInputElement> {
  type?: InputTextTypes
}

export const InputText = styled<InputTextProps, 'input'>('input')`
  border: solid 1px ${props => props.theme.colors.borderColor};
  height: 28px;
  padding: 0 ${themeSpacing.s};
  border-radius: 4px;
  font-size: ${fontSizes[FontRamp.Five]};
`
