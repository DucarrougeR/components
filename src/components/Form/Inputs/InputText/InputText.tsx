import * as React from 'react'
import { Box, BoxProps } from '../../../Box'
import { InputProps } from '../InputProps'

export interface InputTextProps extends BoxProps, InputProps {
  /**
   * Specifies value of the input field.
   */
  value?: string
  /**
   * Displays an example value or short hint to the user. Should not replace a label.
   */
  placeholder?: string
}

export const InputText: React.SFC<InputTextProps> = ({
  validationType,
  ...props
}) => {
  const handleValidationType = () => {
    switch (validationType) {
      case 'error':
        return 'palette.red000'
      default:
        return undefined
    }
  }
  const type = { type: 'text' }
  return (
    <Box
      is="input"
      {...props}
      {...type}
      borderRadius="4px"
      height="28px"
      py="none"
      px="small"
      fontSize="5"
      bg={handleValidationType()}
      border="solid 1px"
      borderColor="palette.charcoal300"
    />
  )
}
