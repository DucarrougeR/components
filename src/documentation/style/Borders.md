<div class="component-desc"><p>Borders help define the space and structure of the interface. </p></div>

<div class="doc-section-divider"></div>

```js noeditor
const BorderRender = require('../../../styleguide_components/BorderRender').BorderRender;

<BorderRender />
```
