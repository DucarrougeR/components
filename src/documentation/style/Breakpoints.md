<div class="component-desc"><p>Breakpoints give hooks for components to adapt and change in relation to the window size.</p></div>

<div class="doc-section-divider"></div>


### Available Breakpoints

```js noeditor
const BreakpointOptionsRender = require('../../../styleguide_components/BreakpointTable').BreakpointOptionsRender;

<BreakpointOptionsRender />
```


