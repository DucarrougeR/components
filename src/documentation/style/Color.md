<div class="component-desc"><p>The Lens color palette aims to help communicate the intent of an interface as well as to help establish clear hierarchy throughout an application. </p></div>

<div class="doc-section-divider"></div>

#### Guidelines

These guidelines lay a few ground rules on how to most effectively use color when creating Looker experiences.

##### **Purposeful**

It is important to use color with intent. Color should support the content and not the other way around. Avoid only using color as a way to show intent.

##### **Respectful**

Use color to create hierarchy and focus attention. Understand that color can have different cultural connotations. Be judicious and use color to draw attention to things the user should take action on.

##### **Accessible**

We want Looker to be accessible to the widest audience possible. Our color palettes are designed to meet or exceed the accessibility guidelines.

<div class="doc-section-divider"></div>

#### Color Palette

```js noeditor
import { Text } from '../../components/Text'

const polished = require('polished')

checkContrast = color => {
  let swatchList = []
  let groupName = color.name
  color.children.forEach((swatch, i) => {
    // Extract swatch later and get fill color
    const swatchLayer = swatch.children.find(function(obj) {
      return obj.name === 'swatch'
    })
    const swatchFill = swatchLayer.fills[0].color
    const fill = polished.rgba(
      swatchFill.r * 255,
      swatchFill.g * 255,
      swatchFill.b * 255,
      swatchFill.a * 1
    )

    // Extract text layer and get color
    const swatchLabel = swatch.children.find(function(obj) {
      return obj.name === 'label'
    })

    const labelFill = swatchLabel.fills[0].color
    const labelColor = polished.rgba(
      labelFill.r * 255,
      labelFill.g * 255,
      labelFill.b * 255,
      labelFill.a * 1
    )

    const hexValue = polished.rgbToColorString({
      red: swatchFill.r * 255,
      green: swatchFill.g * 255,
      blue: swatchFill.b * 255,
    })

    const labelLuminance = polished.getLuminance(labelColor)
    const fillLuminance = polished.getLuminance(fill)
    const constrast =
      (Math.max(labelLuminance, fillLuminance) + 0.05) /
      (Math.min(labelLuminance, fillLuminance) + 0.05)

    const contrastLevel =
      constrast >= 4.5 ? 'AAA' : constrast >= 3 ? 'AA' : 'fail'

    swatchList.push({
      fill,
      label: swatch.name,
      labelColor,
      contrastLevel,
      hexValue,
      group: groupName,
    })
  })

  return swatchList
}

renderSwatch = (swatchList, key) => {
  let color = swatchList[0].group
  let fillColor = swatchList[5].fill
  let fill = color.toLowerCase()
  return (
    <div className="swatch-holder" key={key}>
      <div className="swatch-header" style={{ background: fillColor }}>
        <Text fontSize="large" fontWeight="semiBold" style={{ color: '#fff' }}>
          {color}
        </Text>
        <Text fontSize="medium" fontWeight="semiBold" style={{ color: '#fff' }}>
          500
        </Text>
      </div>
      {swatchList.map((swatch, index) => (
        <Swatch
          key={index}
          contrastLevel={swatch.contrastLevel}
          labelColor={swatch.labelColor}
          hexValue={swatch.hexValue}
          group={swatch.group}
        >
          {swatch.label}
        </Swatch>
      ))}
    </div>
  )
}

const Swatch = props => (
  <div className="swatch" style={{ background: props.hexValue }}>
    <Text
      className="swatch-label"
      fontWeight="semiBold"
      style={{ color: props.labelColor }}
    >
      {props.children}
    </Text>
    <Text
      className="swatch-hex"
      fontWeight="semiBold"
      style={{ color: props.labelColor }}
    >
      {props.hexValue}
    </Text>
    <div className="contrast-box">{props.contrastLevel}</div>
  </div>
)

class SwatchRender extends React.Component {
  constructor() {
    super()
    this.state = {
      file: 'XZUILHEQBKTus121SD99nrFN',
      page: 'Swatches',
      frame: 'contrast',
      ACCESS_TOKEN: '16-4ef5a47f-6cd6-4372-95c3-2d581b1d4bbf',
      swatches: [],
    }
  }

  apiRequest(endpoint) {
    return fetch('https://api.figma.com/v1' + endpoint, {
      method: 'GET',
      headers: { 'x-figma-token': this.state.ACCESS_TOKEN },
    })
      .then(function(response) {
        return response.json()
      })
      .catch(function(error) {
        return { err: error }
      })
  }

  getComponentPage(obj, pageName, id) {
    return obj.find(page => page.name === pageName)
  }

  loadFile() {
    this.apiRequest('/files/' + this.state.file).then(response => {
      let page = this.getComponentPage(
        response.document.children,
        this.state.page
      )
      let pageSwatches = this.getComponentPage(page.children, this.state.frame)
      let swatchList = []
      let palettes = pageSwatches.children

      palettes.forEach(palette => {
        swatchList.push(checkContrast(palette))
      })

      this.setState({
        swatches: swatchList,
      })
    })
  }

  componentDidMount() {
    this.loadFile()
  }

  render() {
    return (
      <div className="swatches">
        {this.state.swatches.map((swatchList, index) =>
          renderSwatch(swatchList, index)
        )}
      </div>
    )
  }
}

;<SwatchRender />
```
