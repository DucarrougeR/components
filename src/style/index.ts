export * from './animations'
export * from './helpers'
export * from './theme'
export * from './styled_components'
