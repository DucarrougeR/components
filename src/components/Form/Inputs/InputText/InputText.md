```js noeditor
const StatusAndResources = require('../../../../../styleguide_components/StatusAndResources')
  .StatusAndResources

;<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/2MG6DoSjk4IaLnjjFCnKFf/Forms"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Form/Inputs/InputText.tsx"
  feedbackTitle="InputText Component Feedback"
/>
```

### Default InputText

The `<InputText />` component renders a single text input element on the page, with no accompanying label. It is generally used to construct higher-order components like the `<FieldText />`. If you are building a form, you probably want to use `<FieldText />` instead as it provides label and validation support.

```js
import { InputText } from './InputText'
;<InputText />
```

### Name and ID

A name and ID can be specified in the `<InputText />` component. Names are important if the input is used in the context of a form, in which case a name is required for the value of the input to be captured.

```js
import { InputText } from './InputText'
;<InputText name="someName" id="someId" />
```

### Disabled Property

Use the disable property to make an input field uneditable.

```js
import { InputText } from './InputText'
;<>
  <InputText disabled />
  <InputText value="Hello!" disabled />
</>
```

### Placeholders

Placeholders are used to give a hint to the user of the expected value for the input. They should not be used as a complete replacement of labels.

```js
import { InputText } from './InputText'
;<InputText placeholder="Enter your text here..." />
```

### ReadOnly Property

As the name suggests, `readOnly` makes the text uneditable.

```js
import { InputText } from './InputText'
;<InputText value="You can't change me." readOnly />
```

### Required Property

When set for `<InputText />` in the context of a form, the input field must be filled out before submitting the form.

```js
import { InputText } from './InputText'
import { Button } from '../../../Button'
;<form>
  <InputText required />
  <br />
  <br />
  <Button>Submit</Button>
</form>
```
