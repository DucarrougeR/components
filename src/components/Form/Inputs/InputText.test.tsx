import { mount } from 'enzyme'
import 'jest-styled-components'
import * as React from 'react'
import { create } from 'react-test-renderer'
import { theme } from '../../../style'
import { InputText } from './InputText'

test('InputText default', () => {
  const component = create(<InputText theme={theme} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('InputText with name and id', () => {
  const component = create(<InputText theme={theme} name="Bob" id="Bobby" />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('InputText should accept disabled', () => {
  const component = create(<InputText theme={theme} disabled />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('InputText with a placeholder', () => {
  const component = create(
    <InputText theme={theme} placeholder="I am a placeholder" />
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('InputText should accept readOnly', () => {
  const component = create(<InputText theme={theme} readOnly />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('InputText should accept required', () => {
  const component = create(<InputText theme={theme} required />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('InputText with a value', () => {
  const component = create(<InputText theme={theme} value="Some value" />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('InputText with aria-describedby', () => {
  const component = create(
    <InputText theme={theme} aria-describedby="some-id" />
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Should trigger onChange handler', () => {
  let counter = 0
  const handleChange = () => counter++

  const wrapper = mount(<InputText theme={theme} onChange={handleChange} />)

  wrapper.find('input').simulate('change', { target: { value: '' } })
  expect(counter).toEqual(1)
})
