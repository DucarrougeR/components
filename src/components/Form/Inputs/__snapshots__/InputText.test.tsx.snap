// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`An InputText 1`] = `
.c0 {
  border: solid 1px #C8CBCC;
  height: 28px;
  padding: 0 0.5rem;
  border-radius: 4px;
  font-size: 0.875rem;
}

<input
  className="c0"
  id="thumbs-up"
  label="👍"
  name="thumbsUp"
/>
`;
